crates/deltalake-core/src/kernel/snapshot/extract.rs

```diff
@@ -0,0 +1,167 @@
+    name: &'a str,
+) -> DeltaResult<&'a T> {
+    extract_and_cast_opt::<T>(arr, name).ok_or(DeltaTableError::Generic(format!(
+        "missing-column: {}",
+        name
+    )))
+}
+
+/// Extracts a column by name and casts it to the given type array type `T`.
+///
+/// Returns `None` if the column does not exist or if the column is not of type `T`.
+pub(super) fn extract_and_cast_opt<'a, T: Array + 'static>(
+    array: &'a dyn ProvidesColumnByName,
+    name: &'a str,
+) -> Option<&'a T> {
+    let mut path_steps = name.split('.');
+    let first = path_steps.next()?;
+    extract_column(array, first, &mut path_steps)
+        .ok()?
+        .as_any()
+        .downcast_ref::<T>()
+}
+
+pub(super) fn extract_column<'a>(
+    array: &'a dyn ProvidesColumnByName,
+    path_step: &str,
+    remaining_path_steps: &mut impl Iterator<Item = &'a str>,
+) -> Result<&'a Arc<dyn Array>, ArrowError> {
+    let child = array
+        .column_by_name(path_step)
+        .ok_or(ArrowError::SchemaError(format!(
+            "No such field: {}",
+            path_step,
+        )))?;
+
+    if let Some(next_path_step) = remaining_path_steps.next() {
+        match child.data_type() {
+            DataType::Map(_, _) => {
+                // NOTE a map has exatly one child, but we wnat to be agnostic of its name.
+                // so we case the current array as map, and use the entries accessor.
+                let maparr = cast_column_as::<MapArray>(path_step, &Some(child))?;
+                if let Some(next_path) = remaining_path_steps.next() {
+                    extract_column(maparr.entries(), next_path, remaining_path_steps)
+                } else {
+                    Ok(child)
+                    // if maparr.entries().num_columns() != 2 {
+                    //     return Err(ArrowError::SchemaError(format!(
+                    //         "Map {} has {} columns, expected 2",
+                    //         path_step,
+                    //         maparr.entries().num_columns()
+                    //     )));
+                    // }
+                    // if next_path_step == *maparr.entries().column_names().first().unwrap() {
+                    //     Ok(maparr.entries().column(0))
+                    // } else {
+                    //     Ok(maparr.entries().column(1))
+                    // }
+                }
+            }
+            DataType::List(_) => {
+                let listarr = cast_column_as::<ListArray>(path_step, &Some(child))?;
+                if let Some(next_path) = remaining_path_steps.next() {
+                    extract_column(
+                        cast_column_as::<StructArray>(next_path_step, &Some(listarr.values()))?,
+                        next_path,
+                        remaining_path_steps,
+                    )
+                } else {
+                    Ok(listarr.values())
+                }
+            }
+            _ => extract_column(
```

Comment:
Won't this fail when you have a field that is called random.field and the the actual datatype is a non-nested dtype? Maybe we should the if `dataype.is_nested()` first

Reply:
`.` is only allowed in column names if collumn mapping is enabled. Since we do not yet support that feature, I wanted to leave this for a follow up ...